File: shell.c
#include <stdio.h>

void print(int arr[], int cnt)
{
    int i;
    for (i = 0; i < cnt; i++)
    {
        printf(" %d ", arr[i]);
    }
    printf("\n");
}

void shell_sort(int arr[], int cnt)
{
    int gap = 0;
    while (gap <= cnt)
    {
        gap = gap * 3 + 1;
    }
    printf("%d\n", gap);
    int i, k, temp;
    while (gap > 0)
    {
        printf("------------\n");
        for (i = gap; i < cnt; i++)
        {
            printf("============\n");
            temp = arr[i];
            k = i - gap;
            while ((k >= 0) && (arr[k] > temp))
            {
                arr[k + gap] = arr[k];
                k = k - gap;
            }
            arr[k + gap] = temp;
        }
        gap = (gap - 1) / 3;
    }
}

int main(void)
{
    int arr[] = {5, 4, 3, 2, 1};
    const int cnt = 5;
    print(arr, cnt);
    shell_sort(arr, cnt);
    print(arr, cnt);
}

File: shell.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from random import randint
import copy

def prepare():
    cnt = 10
    return [randint(0, cnt) for _ in range(cnt)]

def swap(arr, i, j):
    arr[i], arr[j] = arr[j], arr[i]

def shell_sort(arr):
    length = len(arr)
    step = 1
    while step < length / 3:
        step = 3 * step + 1
    while step >= 1:
        for i in range(step):
            for j in range(i + step, length, step):
                for k in range(j, i, -(step)):
                    if arr[k] < arr[k - step]:
                        swap(arr, k - step, k)
                    else:
                        break
        step /= 3

def shell_sort2(arr):
    length = len(arr)
    step = 1
    while step < length / 3:
        step = 3 * step + 1
    while step >= 1:
        for i in range(step, length):
            for j in range(i, 0, -(step)):
                if arr[j] < arr[j - step]:
                    swap(arr, j - step, j)
                else:
                    break
        step /= 3

def main():
    arr = prepare()
    arr2 = copy.deepcopy(arr)
    print arr
    shell_sort(arr)
    print arr
    print arr2
    shell_sort2(arr2)
    print arr2

if __name__ == '__main__':
    main()
